### ISSUE TITLE
On Android 11 mobile phone, I encountered an error when initializing the database.

### ISSUE THREAD

**Issue by kiddik666** [powersync-ja/sqlite_async.dart#42]
05-15 06:31:23.345   497  1813 I ActivityTaskManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER
] flg=0x10200000 cmp=com.example.ji_zhang/.MainActivity bnds=[37,1236][238,1560]} from uid 10152
05-15 06:31:23.369   497   529 I ActivityManager: Start proc 4486:com.example.ji_zhang/u0a171 for pre-top-activity {com.example.ji_z
hang/com.example.ji_zhang.MainActivity}
05-15 06:31:23.961  4486  4486 W xample.ji_zhang: type=1400 audit(0.0:41): avc: denied { read } for name="max_map_count" dev="proc"
ino=84487 scontext=u:r:untrusted_app:s0:c171,c256,c512,c768 tcontext=u:object_r:proc_max_map_count:s0 tclass=file permissive=0 app=c
om.example.ji_zhang

**Comment by rkistner** [comment]
Those logs appear to display a warning message, but no indication of a hard error or crash. What are the symptoms you're seeing? And does it work fine on other devices?

**Comment by kiddik666** [comment]
> Those logs appear to display a warning message, but no indication of a hard error or crash. What are the symptoms you're seeing? And does it work fine on other devices?

After starting the program, the elements in the main interface cannot be displayed and  keeps staying on the program’s startup screen.    If I delete this line of code(await AccountingLogic.database;)  ,My program can enter the main interface normally and display various interface elements.

 The errors displayed in logcat are the ones I posted above.



 
import 'package:flutter/material.dart';
import 'package:permission_handler/permission_handler.dart';

import 'package:flutter/services.dart';
import 'package:path/path.dart' as pth;
import 'dart:ffi';
import 'dart:io';



 void requestManageStoragePermission() async {
     var status = await Permission.manageExternalStorage.status;
     if (!status.isGranted) {
       openAppSettings(); 
     }
   }


void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  requestManageStoragePermission();
  
  //If I delete this line of code, I can see all the interface elements on the main screen after launching the program, otherwise it will stay on the startup screen indefinitely.
  await AccountingLogic.database;

  runApp(MyApp());
}





import 'package:sqlite_async/sqlite_async.dart';
import 'package:sqlite3_flutter_libs/sqlite3_flutter_libs.dart';


class AccountingLogic {
  static late SqliteDatabase _database;
  static final _migrations = SqliteMigrations()
    ..add(SqliteMigration(1, (tx) async {
      await tx.execute('''
        CREATE TABLE IF NOT EXISTS accounting (
          id INTEGER PRIMARY KEY,
          year INTEGER,
          month INTEGER,
          roomNumber INTEGER,
          rentPaid REAL,
          electricityPaid REAL,
          propertyPaid REAL,
          modificationLog TEXT,
          remark TEXT
        )
      ''');
      await tx.execute('''
        CREATE UNIQUE INDEX IF NOT EXISTS idx_unique_record ON accounting (year, month, roomNumber)
      ''');
    }));
    

  static Future<SqliteDatabase> get database async {
    if (_database == null) {
      _database = SqliteDatabase(path: '/storage/emulated/0/Documents/accounting.db');
      await _migrations.migrate(_database);
    }
    return _database;
  }
  
  ......
  ......
  
  }

**Comment by rkistner** [comment]
The issue may be related to permissions, so try with an app-specific path first (you don't need any permissions for this):

```dart
import 'package:path_provider/path_provider.dart';
...
final dir = await getApplicationSupportDirectory();
final path = join(dir.path, 'accounting.db');
_database = SqliteDatabase(path: path);
```

If that doesn't work, the issue is likely something with the migrations, and you should try use a debugger to find the issue. If it does work, you know your current issue is likely due to the specific location and/or permissions.

**Comment by kiddik666** [comment]
> The issue may be related to permissions, so try with an app-specific path first (you don't need any permissions for this):
> 
> ```dart
> import 'package:path_provider/path_provider.dart';
> ...
> final dir = await getApplicationSupportDirectory();
> final path = join(dir.path, 'accounting.db');
> _database = SqliteDatabase(path: path);
> ```
> 
> If that doesn't work, the issue is likely something with the migrations, and you should try use a debugger to find the issue. If it does work, you know your current issue is likely due to the specific location and/or permissions.

I solved this problem:
I generated my program as a Windows application, and after executing `flutter run -d windows`, it prompted: "Cannot use it before _database initialization", indicating that the problem was with the line `if (_database == null)`. I deleted this line of code and the problem was resolved.This shows that _database cannot detect whether it is null before initialization.
This also indicates that using (adb logcat | Select-String -Pattern 'com.example.ji_zhang') cannot display all error messages. Moreover, the error message I posted earlier does not affect the normal operation of the program.
I cannot determine whether Logcat in Android Studio can display all error messages, as my Android Studio keeps showing 'connecting to the emulator' and cannot start successfully .
As it turns out, having a tool that can display the cause of errors is very important, otherwise, it will waste a lot of time.

At the same time, thank you for your enthusiasm to help me.
